### ISSUE TITLE
Add optional gradient tensor fields to training response types for client-side analysis

### ISSUE THREAD

**Issue by nshkrdotcom** [thinking-machines-lab/tinker-feedback#28]
## Summary
Extend ForwardBackwardOutput and related training response types to optionally include per-parameter gradient tensors, enabling researchers to implement client-side analysis workflows for Fisher Information Matrix approximation, adaptive rank reduction, and custom convergence telemetry while preserving Tinker's client-driven training architecture.

## Current State and Motivation

Tinker's training API centers on client-driven execution through primitives like forward_backward and optim_step exposed by TrainingClient. This architecture gives researchers fine-grained control over training loops while Tinker's infrastructure handles the computational heavy lifting of forward passes, gradient computation, and parameter updates. The current response types (ForwardBackwardOutput, OptimStepResponse) provide scalar metrics and loss function outputs but do not include gradient tensors themselves.

This design choice makes sense for standard training workflows where aggregate statistics like global gradient norms provide sufficient observability. However, advanced research in model behavior and training dynamics increasingly requires access to gradient tensor information at a finer granularity than summary statistics. The Jacobian analysis framework demonstrates several research capabilities that become possible with gradient tensor access but remain inaccessible through current observability mechanisms.

Researchers studying neural network training dynamics need to observe how gradients flow through different components of the model to identify bottlenecks, validate theoretical predictions about convergence behavior, and understand how architectural choices affect learning patterns. Projects focused on parameter-efficient fine-tuning need data-driven methods for selecting LoRA ranks and determining which layers most benefit from adaptation. Work on model interpretability requires tracing how input features influence parameter updates through gradient attribution techniques. Research on adversarial robustness needs to analyze gradient patterns to develop defenses or understand attack surfaces.

The current workarounds for these research needs involve significant compromises. Some researchers implement offline analysis by saving checkpoints at frequent intervals and estimating gradients through parameter differences, but this approach suffers from both inaccuracy (numerical differentiation artifacts) and inefficiency (checkpoint storage and loading overhead). Others abandon managed training platforms entirely to implement custom training loops from scratch, sacrificing the reliability, performance optimizations, and operational convenience that motivated choosing a managed platform in the first place. Some resort to approximate techniques like finite difference methods applied to forward passes, which are computationally expensive and numerically unstable. None of these alternatives provide the combination of precision, efficiency, and convenience that direct gradient access would enable.

### Core API Extension

Extend ForwardBackwardOutput to include an optional gradients field containing per-parameter gradient tensors:

```python
class ForwardBackwardOutput(BaseModel):
    loss_fn_output_type: str
    loss_fn_outputs: List[LossFnOutput]
    metrics: Dict[str, float]
    gradients: Optional[Dict[str, TensorData]] = None  # New field
```

Add a configuration parameter to ForwardBackwardRequest allowing clients to opt into gradient returns with controllable detail level:

```python
forward_backward_request = ForwardBackwardRequest(
    data=batch_data,
    loss_fn="cross_entropy",
    return_gradients="lora_only",  # Options: None, "lora_only", "full", "sampled"
)
```

### Client-Side Execution Pattern

Gradient analysis would execute entirely in client code, maintaining Tinker's existing architecture where the client orchestrates training steps. A typical research workflow would follow this pattern:

```python
def train_with_gradient_analysis(training_client, dataloader, gradient_callback, callback_frequency=10):
    for step, batch in enumerate(dataloader):
        # Standard forward-backward with gradient returns every N steps
        return_grads = "lora_only" if step % callback_frequency == 0 else None
        
        fwdbwd_future = training_client.forward_backward(
            batch, 
            "cross_entropy",
            return_gradients=return_grads
        )
        fwdbwd_result = fwdbwd_future.result()
        
        # Client-side gradient analysis when available
        if fwdbwd_result.gradients is not None:
            step_info = {"step": step, "loss": fwdbwd_result.metrics.get("loss")}
            analysis_metrics = gradient_callback(fwdbwd_result.gradients, step_info)
            # Analysis metrics can be logged alongside training metrics
        
        # Continue with optimizer step as normal
        optim_future = training_client.optim_step(adam_params)
        optim_future.result()
```

The gradient callback executes in the client process where researchers have full control over the execution environment, can import arbitrary analysis libraries, and can implement computationally intensive operations without impacting Tinker's training infrastructure.

### Gradient Detail Levels

The return_gradients parameter would support multiple detail levels to balance research needs against bandwidth and computational costs:

The "lora_only" mode returns gradients only for LoRA adapter parameters, dramatically reducing data transfer volume while still enabling the most common analysis workflows for parameter-efficient fine-tuning research. This mode would be the recommended default for most research applications.

The "full" mode returns gradients for all trainable parameters in the model. This option supports comprehensive analysis but requires careful consideration of bandwidth costs and should typically be combined with infrequent callback execution.

The "sampled" mode returns gradients for a random subset of parameters or layers, enabling statistically valid analysis while controlling data transfer volume. The sampling strategy could be configured through additional parameters specifying layer patterns or sampling rates.

An explicit None or absence of the parameter maintains current behavior where no gradients are returned, ensuring backward compatibility and zero overhead for training workflows that do not require gradient analysis.

### Bandwidth and Performance

Shipping gradient tensors over network connections introduces substantial bandwidth costs that must be carefully managed. For a typical LoRA configuration with rank sixteen adapters across multiple layers, gradient tensors can total hundreds of megabytes per training step. The combination of detail level control and callback frequency parameters gives researchers explicit tools for managing this tradeoff. Documentation should include clear guidance on bandwidth implications and recommended callback frequencies for different model sizes and network conditions.

The implementation should leverage existing TensorData serialization mechanisms and consider compression strategies for gradient tensors. Since gradients often exhibit sparsity or low-rank structure, modest compression could significantly reduce transfer costs without meaningful information loss for analysis purposes.

### Parameter Naming and Stability

Exposing gradient tensors requires establishing a public, stable naming convention for model parameters. The current API provides model-level metadata like lora_rank in training run descriptions but does not expose individual parameter names. The implementation must define a consistent naming scheme (for example, "blocks.3.attn.q_proj.lora_A") and commit to maintaining this naming stability across API versions so that analysis code remains reliable across training runs.

Parameter naming should follow established conventions from the underlying model architectures when possible, making it easier for researchers to map parameter names to architectural components. Clear documentation should specify the naming scheme and provide examples showing how to locate specific components of interest.

### Type Consistency

The proposed API additions maintain consistency with Tinker's existing type system built around Pydantic models and TensorData. Gradient tensors would be represented as TensorData objects in the response, allowing seamless integration with existing client code patterns. Client-side analysis callbacks can convert TensorData to framework-specific tensor types (PyTorch tensors, NumPy arrays) as needed for their analysis libraries.

### Memory and Storage

Client applications implementing gradient callbacks must manage memory carefully, particularly when accumulating gradient information across multiple steps for techniques like Fisher Information Matrix approximation. The implementation should provide clear guidance on memory management patterns and potentially offer utility functions for efficient gradient accumulation and aggregation.

### Fisher Information Matrix Approximation

Researchers can build FIM approximations by accumulating outer products of gradient vectors across training steps:

```python
class FisherInformationTracker:
    def __init__(self):
        self.fisher_diagonal = {}
    
    def gradient_callback(self, gradients, step_info):
        for param_name, grad_tensor_data in gradients.items():
            grad = torch.tensor(grad_tensor_data.data)
            if param_name not in self.fisher_diagonal:
                self.fisher_diagonal[param_name] = torch.zeros_like(grad)
            self.fisher_diagonal[param_name] += grad ** 2
        
        return {"fim_trace": sum(f.sum().item() for f in self.fisher_diagonal.values())}
```

This pattern enables principled capacity allocation decisions based on information geometry rather than heuristics.

### Adaptive Rank Monitoring

Researchers can monitor whether configured LoRA ranks match the effective dimensionality of adaptations by analyzing gradient singular values:

```python
def rank_analysis_callback(gradients, step_info):
    metrics = {}
    for param_name, grad_tensor_data in gradients.items():
        if "lora" in param_name:
            grad = torch.tensor(grad_tensor_data.data)
            if grad.ndim == 2:  # Weight matrix
                singular_values = torch.linalg.svdvals(grad)
                effective_rank = (singular_values > 0.01 * singular_values[0]).sum()
                metrics[f"{param_name}_effective_rank"] = effective_rank.item()
    return metrics
```

This analysis reveals when rank reduction could improve efficiency without sacrificing quality, or when rank increases might improve model capacity.

### Convergence Telemetry

Researchers can implement custom convergence metrics that standard training systems do not provide:

```python
def convergence_analysis_callback(gradients, step_info):
    gradient_magnitudes = [
        torch.tensor(g.data).norm().item() 
        for g in gradients.values()
    ]
    
    return {
        "max_grad_norm": max(gradient_magnitudes),
        "grad_norm_variance": torch.tensor(gradient_magnitudes).var().item(),
        "grad_norm_90th_percentile": torch.quantile(
            torch.tensor(gradient_magnitudes), 0.9
        ).item()
    }
```

These metrics help diagnose training instabilities, identify learning rate issues, or validate theoretical predictions about gradient flow.

## Benefits to Research Community

Direct gradient access transforms Tinker from a platform for standard training into a platform for training dynamics research. Projects that previously required custom training implementations to study optimization behavior, parameter sensitivity, or learning dynamics can now leverage Tinker's managed infrastructure while maintaining access to the gradient information their research requires.

The client-side execution model preserves researcher autonomy. Analysis code runs in environments researchers control, where they can install specialized libraries, implement computationally intensive algorithms, and integrate gradient analysis with external tools. This flexibility enables research workflows that would be impossible if gradient analysis had to execute within Tinker's infrastructure.

The optional, configurable nature of gradient returns ensures zero overhead for workflows that do not need this capability while making it accessible for those that do. Researchers can enable gradient callbacks during exploratory phases when they need deep visibility into training dynamics, then disable them for production training runs when performance is paramount.

## Relationship to Existing Architecture

This proposal respects and builds upon Tinker's fundamental design principles. Training remains client-driven with explicit control over forward passes, backward passes, and optimizer steps. Tinker infrastructure handles the computationally intensive work of gradient computation but does not prescribe how clients use that information. The client-side callback pattern maintains clear separation between managed training primitives and research-specific analysis.

The implementation would be purely additive. Existing training code would continue to work without modification. The new gradient fields would remain None unless explicitly requested, ensuring backward compatibility. The enhancement extends Tinker's value proposition for advanced research use cases without compromising the simplicity and reliability that make it effective for standard training workflows.

By providing gradient tensor access through response type extensions rather than attempting to integrate arbitrary callbacks into the training loop, this design maintains Tinker's architectural clarity while enabling sophisticated research workflows that current observability mechanisms cannot support. The approach positions Tinker as infrastructure that provides powerful primitives while leaving research-specific logic firmly in client control.
